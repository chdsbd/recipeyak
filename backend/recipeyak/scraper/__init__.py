"""
Fetch recipe data from a url and try and avoid SSRF

rel: https://github.com/getsentry/sentry/blob/66b93770e95290a3ab257311e4a2598304fb4e6f/src/sentry/http.py#L57
rel: https://forum.sentry.io/t/source-code-scrapping/4167/7
"""
from __future__ import annotations

import time
from dataclasses import asdict, dataclass
from io import BytesIO

from django.core.validators import URLValidator
from recipe_scrapers import scrape_html
from typing_extensions import TypedDict
from urllib3.util.retry import Retry
from yarl import URL

from recipeyak.models import Scrape
from recipeyak.scraper.safe_session import SafeSession


class Review(TypedDict):
    author: str
    datePublished: str
    reviewRating: float
    reviewBody: str


@dataclass
class ScrapeResult:
    title: str | None
    total_time: str | None
    yields: str | None
    image: str | None
    ingredients: list[str]
    instructions: list[str]
    author: str | None
    canonical_url: str | None
    # db id
    id: int | None = None


MAX_RES_LENGTH = 40 * 1024 * 1024  # 40MB
TIMEOUT = 5

TIME_DURATION_UNITS = (
    ("week", 60 * 60 * 24 * 7),
    ("day", 60 * 60 * 24),
    ("hour", 60 * 60),
    ("min", 60),
    ("sec", 1),
)


def human_time_duration(seconds: int) -> str:
    """
    from: https://gist.github.com/borgstrom/936ca741e885a1438c374824efb038b3
    """
    if seconds == 0:
        return "inf"
    parts = []
    for unit, div in TIME_DURATION_UNITS:
        amount, seconds = divmod(int(seconds), div)
        if amount > 0:
            parts.append("{} {}{}".format(amount, unit, "" if amount == 1 else "s"))
    return ", ".join(parts)


def validate_url(url: str) -> None:
    URLValidator(schemes=["https", "http"])(url)


def scrape_recipe(*, url: str) -> ScrapeResult:
    """
    fetch a recipe and avoid:
    - fetching a really large file
    - taking forever to fetch a file
    - hitting an internal IP (aka SSRF)
    """

    if not url.startswith("http"):
        url = "https://" + url

    validate_url(url)

    start = time.monotonic()

    with SafeSession(
        max_retries=Retry(
            total=3, backoff_factor=0.3, status_forcelist=(429, 500, 502, 503, 504)
        )
    ) as http:
        response = http.get(
            url,
            timeout=TIMEOUT,
            allow_redirects=True,
            stream=True,
            headers={
                # naive attempt to look like a browser
                "Host": URL(url).host,
                "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.5 Safari/605.1.15",
                "Accept-Language": "en-US,en;q=0.9",
                "Connection": "keep-alive",
            },
        )
    response.raise_for_status()

    # via https://stackoverflow.com/a/22347526
    # and https://github.com/getsentry/sentry/blob/66b93770e95290a3ab257311e4a2598304fb4e6f/src/sentry/http.py#L171
    try:
        content_len = int(response.headers["content-length"])
    except (LookupError, ValueError):
        content_len = 0
    if content_len > MAX_RES_LENGTH:
        raise OverflowError
    buf = BytesIO()
    size = 0
    for chunk in response.iter_content(16 * 1024):
        if time.monotonic() - start > TIMEOUT:
            raise TimeoutError
        buf.write(chunk)
        size += len(chunk)
        if size > MAX_RES_LENGTH:
            raise OverflowError

    end = time.monotonic()

    page_data = buf.getvalue()

    r = scrape_html(html=page_data, org_url=url)

    scrape_result = ScrapeResult(
        canonical_url=r.canonical_url(),
        title=r.title(),
        total_time=human_time_duration(r.total_time() * 60),
        yields=r.yields(),
        image=r.image(),
        ingredients=r.ingredients(),
        instructions=r.instructions_list(),
        author=r.author(),
    )
    parsed = asdict(scrape_result)
    del parsed["id"]
    scrape = Scrape.objects.create(
        html=page_data,
        url=url,
        duration_sec=end - start,
        parsed=parsed,
    )
    scrape_result.id = scrape.id

    return scrape_result
