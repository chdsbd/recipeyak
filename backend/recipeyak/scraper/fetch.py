import time
from io import BytesIO

from urllib3.util.retry import Retry
from yarl import URL

from recipeyak.scraper.safe_session import SafeSession

MAX_RES_LENGTH = 40 * 1024 * 1024  # 40MB
TIMEOUT = 5


def fetch_bytes(*, url: str) -> bytes:
    start = time.monotonic()
    with SafeSession(
        max_retries=Retry(
            total=3, backoff_factor=0.3, status_forcelist=(429, 500, 502, 503, 504)
        )
    ) as http:
        response = http.get(
            url,
            timeout=TIMEOUT,
            allow_redirects=True,
            stream=True,
            headers={
                # naive attempt to look like a browser
                "Host": URL(url).host,
                "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.5 Safari/605.1.15",
                "Accept-Language": "en-US,en;q=0.9",
                "Connection": "keep-alive",
            },
        )
    response.raise_for_status()

    # via https://stackoverflow.com/a/22347526
    # and https://github.com/getsentry/sentry/blob/66b93770e95290a3ab257311e4a2598304fb4e6f/src/sentry/http.py#L171
    try:
        content_len = int(response.headers["content-length"])
    except (LookupError, ValueError):
        content_len = 0
    if content_len > MAX_RES_LENGTH:
        raise OverflowError
    buf = BytesIO()
    size = 0
    for chunk in response.iter_content(16 * 1024):
        if time.monotonic() - start > TIMEOUT:
            raise TimeoutError
        buf.write(chunk)
        size += len(chunk)
        if size > MAX_RES_LENGTH:
            raise OverflowError
    return buf.getvalue()


def fetch_content_length(*, url: str) -> int | None:
    """
    grab the content-length of the response without downloading the entire file
    """
    with SafeSession(
        max_retries=Retry(
            total=3, backoff_factor=0.3, status_forcelist=(429, 500, 502, 503, 504)
        )
    ) as http:
        response = http.get(
            url,
            timeout=TIMEOUT,
            allow_redirects=True,
            stream=True,
            headers={
                # naive attempt to look like a browser
                "Host": URL(url).host,
                "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.5 Safari/605.1.15",
                "Accept-Language": "en-US,en;q=0.9",
                "Connection": "keep-alive",
            },
        )
    response.raise_for_status()

    # via https://stackoverflow.com/a/22347526
    # and https://github.com/getsentry/sentry/blob/66b93770e95290a3ab257311e4a2598304fb4e6f/src/sentry/http.py#L171
    try:
        size = int(response.headers["content-length"])
        if size > MAX_RES_LENGTH:
            raise OverflowError
        return size
    except (LookupError, ValueError):
        return None
