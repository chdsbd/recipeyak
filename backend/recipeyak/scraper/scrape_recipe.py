"""
Fetch recipe data from a url and try and avoid SSRF

rel: https://github.com/getsentry/sentry/blob/66b93770e95290a3ab257311e4a2598304fb4e6f/src/sentry/http.py#L57
rel: https://forum.sentry.io/t/source-code-scrapping/4167/7
"""

from __future__ import annotations

import time
from collections.abc import Sequence
from dataclasses import asdict, dataclass
from pathlib import Path
from urllib.parse import urlparse
from uuid import uuid4

from django.core.exceptions import ValidationError
from django.core.validators import URLValidator
from recipe_scrapers import scrape_html as parse_html

from recipeyak import config
from recipeyak.models import Scrape
from recipeyak.models.upload import Upload
from recipeyak.models.user import User
from recipeyak.scraper.extract import extract_recipe
from recipeyak.scraper.fetch import fetch_bytes, fetch_content_length
from recipeyak.storage import s3


@dataclass
class IngredientGroup:
    name: str | None
    ingredients: list[str]


@dataclass
class ScrapeResult:
    title: str | None
    total_time: str | None
    yields: str | None
    image: str | None
    upload_id: int | None
    instructions: list[str]
    # the first param of the tuple will be None in the case of a
    ingredient_groups: list[IngredientGroup]
    author: str | None
    canonical_url: str | None
    # db id
    id: int | None = None


def _validate_url(url: str) -> None:
    URLValidator(schemes=["https", "http"])(url)


def _parse_url(url: str | None) -> None | str:
    if url is None:
        return None
    try:
        URLValidator(schemes=["https", "http"])(url)
        return url
    except ValidationError:
        return None


def _get_largest_image(urls: Sequence[str | None]) -> str | None:
    """
    fetch the content-length for all the urls and pick the largest one
    """
    max_size: int | None = None
    max_url: str | None = None
    for url in urls:
        url = _parse_url(url)
        if url is None:
            continue
        content_length = fetch_content_length(url=url)
        if content_length is None:
            continue
        if max_size is None or content_length > max_size:
            max_size = content_length
            max_url = url
    return max_url


@dataclass(frozen=True, slots=True, kw_only=True)
class _FindImageResult:
    upload_id: int | None
    image_url: str | None


def _find_and_save_largest_image(
    *, image_urls: Sequence[str | None], user: User
) -> _FindImageResult:
    image_url = _get_largest_image(image_urls)

    upload: Upload | None = None
    if image_url is not None:
        image_res, content_type = fetch_bytes(url=image_url)
        extension = Path(urlparse(image_url).path).suffix
        key = f"scraper/{uuid4().hex}{extension}"

        upload = Upload(
            scraped_by=user,
            bucket=config.STORAGE_BUCKET_NAME,
            key=key,
            content_type=content_type,
        )
        upload.save()
        # put_object seems simplier than upload_fileobj so going with it
        s3().put_object(
            Bucket=config.STORAGE_BUCKET_NAME,
            Key=key,
            Body=image_res,
        )
        upload.completed = True
        upload.save()

    return _FindImageResult(
        upload_id=upload.id if upload is not None else None, image_url=image_url
    )


def scrape_recipe(*, url: str, user: User) -> ScrapeResult:
    """
    fetch a recipe and avoid:
    - fetching a really large file
    - taking forever to fetch a file
    - hitting an internal IP (aka SSRF)
    """
    if not url.startswith("http"):
        url = "https://" + url
    start = time.monotonic()

    _validate_url(url)
    html, _ = fetch_bytes(url=url)

    parsed = parse_html(html=html, org_url=url)  # type: ignore[arg-type]

    recipe = extract_recipe(parsed)

    image = _find_and_save_largest_image(image_urls=recipe.image_urls, user=user)

    scrape_result = ScrapeResult(
        canonical_url=recipe.canonical_url,
        title=recipe.title,
        total_time=recipe.total_time,
        yields=recipe.yields,
        image=image.image_url,
        ingredient_groups=[
            IngredientGroup(name=x.name, ingredients=x.ingredients)
            for x in recipe.ingredient_groups
        ],
        instructions=recipe.instructions,
        author=recipe.author,
        upload_id=image.upload_id,
    )
    scrape_dict = asdict(scrape_result)
    del scrape_dict["id"]
    duration_sec = time.monotonic() - start
    scrape = Scrape.objects.create(
        html=html.decode(),
        url=url,
        duration_sec=duration_sec,
        parsed=scrape_dict,
    )
    scrape_result.id = scrape.id

    return scrape_result
