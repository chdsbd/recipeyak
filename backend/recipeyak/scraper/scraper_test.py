from io import StringIO
from pathlib import Path
from pprint import pprint

from syrupy.assertion import SnapshotAssertion

from recipeyak.scraper.scrape_recipe import _parse_recipe


def test_parse_recipe_tips_html_to_markdown(snapshot: SnapshotAssertion) -> None:
    """
    Check that we convert tips from html to markdown.
    """
    html = (
        Path(__file__).parent
        / "test_data"
        / "1023609-chile-crisp-fettuccine-alfredo-with-spinach.html"
    ).read_bytes()

    result = _parse_recipe(
        html=html,
        url="https://cooking.nytimes.com/recipes/1023609-chile-crisp-fettuccine-alfredo-with-spinach",
    )
    assert result == snapshot()


def test_parse_links_in_steps(snapshot: SnapshotAssertion) -> None:
    """
    Check that we convert links from html to markdown in steps.
    """
    html = (
        Path(__file__).parent
        / "test_data"
        / "pressure-cooker-fast-and-easy-chicken-enchiladas-food-lab-recipe.html"
    ).read_bytes()

    result = _parse_recipe(
        html=html,
        url="https://www.seriouseats.com/pressure-cooker-fast-and-easy-chicken-enchiladas-food-lab-recipe",
    )
    assert result == snapshot()
