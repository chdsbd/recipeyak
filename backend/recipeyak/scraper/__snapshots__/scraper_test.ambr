# name: test_parse_recipe_tips_html_to_markdown
  _ExtractedRecipe(title='Chile Crisp Fettuccine Alfredo With Spinach', total_time='25 mins', yields='6 servings', instructions=['Bring a large pot of salted water to a boil. While the water heats, melt the butter with the chile crisp in a very large skillet or Dutch oven over low heat.', 'Whisk in the cream and keep warm over low. (It should steam, not bubble.)', 'Cook the fettuccine until al dente according to the package directions. Use tongs to transfer the noodles to the cream mixture, reserving the pasta water.', 'Add the spinach and turn with tongs until the noodles are well coated.', 'Add the Parmesan and toss, still over low heat, until the noodles are slicked with a creamy sauce, adding a spoonful or two of pasta water if needed to loosen the sauce. Divide among serving dishes and top with Parmesan and more chile crisp, if youâ€™d like. Serve immediately.', '**Tips**\nYou can make [chile crisp](https://cooking.nytimes.com/recipes/1022366-chile-crisp) easily at home or buy it in supermarkets or online. It varies in spiciness, so adjust the amount to your taste. For this dish, try to add more of the solids than the oil to the sauce for the most flavorful dish.'], ingredient_groups=[_IngredientGroup(name=None, ingredients=['Salt', '4 tablespoons butter', '1 to 2 tablespoons chile crisp, plus more to taste (see Tip)', '1 cup heavy cream', '1 pound dried fettuccine', '1 (5-ounce) package baby spinach', '3/4 cup finely grated Parmesan (2 1/4 ounces), plus more for serving'])], author='Genevieve Ko', canonical_url='https://cooking.nytimes.com/recipes/1023609-chile-crisp-fettuccine-alfredo-with-spinach', image_urls=['https://static01.nyt.com/images/2022/11/02/dining/27korex-alfredo/26korex-alfredo-threeByTwoMediumAt2X.jpg', 'https://static01.nyt.com/images/2022/11/02/dining/27korex-alfredo/27korex-alfredo-square640-v2.jpg'])
# ---
